### ISSUE TITLE
[spanner] Encapsulate enumerator GRPC::BadStatus errors

### ISSUE THREAD

**Issue by thiagotnunes** [googleapis/gapic-generator-ruby#509]
TLDR: we would like to have the conversion of `GRPC::BadStatus` errors to `Google::Cloud::Error`s in the `Enumerable` result of streaming calls of the spanner client, so that the consumer only has to deal with one kind of error.

We are currently observing a problem when calling the streaming calls in the spanner client `Google::Cloud::Spanner::V1::Client` where we have to handle both `Google::Cloud::Error`s and `GRPC::Error`s.

The generated client is rescuing `GRPC::BadStatus` and converting those to `Google::Cloud::Error`s (like [here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L912)). That is fine, because we can rely that if we get a `GRPC::BadStatus` it will be automatically converted to us when executing the method.

The problem lies in handling the result of streaming calls. The generated client is returning `Enumerable<PartialResultSet>` (as specified [here](https://github.com/googleapis/google-cloud-ruby/blob/master/google-cloud-spanner-v1/lib/google/cloud/spanner/v1/spanner/client.rb#L912)). If we call `result.next` we might get `GRPC::BadStatus` errors, because in this stream of results there is no conversion from `GRPC::BadStatus` errors to `Google::Cloud::Error`s. This makes it a bit clunky to deal with the errors, where we have to deal with both flavours of the same errors.

This problem is exemplified below:

```ruby
require "google/cloud/spanner/v1"

client = V1::Spanner::Client.new

begin
  result = client.execute_streaming_sql request, opts # This can raise a Google::Cloud::Error
  result.next # This can raise a GRPC::BadStatus (or other GRPC errors)
rescue GRPC::Unavailable, Google::Cloud::UnavailableError => err # Must rescue both
  puts err
end
```

We have identified this problem when fixing long PDML transactions in the ruby spanner client library (https://github.com/googleapis/google-cloud-ruby/pull/7592).

**Comment by thiagotnunes** [comment]
cc @dazuma

**Comment by viacheslav-rostovtsev** [comment]
So if I understand correctly, the conversion from :GRPC::BadStatus to ::Google::Cloud::Error happens for the regular requests but not for the paginated requests.
This is probably the place: https://github.com/googleapis/gapic-generator-ruby/blob/1b1aaec053b676a190a8433f97b7a43dc66f5673/gapic-common/lib/gapic/paged_enumerable.rb#L132

**Comment by viacheslav-rostovtsev** [comment]
btw @thiagotnunes in your issue text the second 'here' hyperlink (for "The generated client is returning Enumerable<PartialResultSet> as specified here") is the same as first and probably should be different?

**Comment by thiagotnunes** [comment]
@viacheslav-rostovtsev thanks for taking a look at this.

I think your understanding is correct. Thus, if we always converted from `::GRPC::BadStatus` to `::Google::Cloud::Error` it would simplify things in the consumers.

Thanks for pointing the wrong link. I fixed it to point to the correct block of code.
